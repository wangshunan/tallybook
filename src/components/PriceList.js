import React from 'react'
import Ionicon from 'react-ionicons'
import PropTypes from 'prop-types'

const PriceList = ({items, onModifyItem, onDeleteItem}) => {
    return (
        <ul className="list-group">
            {
                items.map((item) => (
                    <li className="list-group-item d-flex justify-content-between align-items-center row" key={item.id} style={{margin:'0px'}}>
                        <span className="col-1 badege badege-primary">
                            <Ionicon
                                className="rounded"
                                fontSize="30px"
                                style={{backgroundColor: '#007bff', padding:'5px'}}
                                color={'#fff'}
                                icon={item.category.iconName}
                            />
                        </span>
                        <span className="col-6">{item.title}</span>
                        <span className="col-1">{(item.category.type === 'income'? '+' : '-') + item.price}</span>
                        <span className="col-2">{item.date}</span>
                        <a className="col-1" role="button" onClick={(e) => {e.preventDefault(); onModifyItem(item.id)}}>
                            <Ionicon
                                className="rounded"
                                fontSize="30px"
                                style={{backgroundColor: '#28a745', padding:'5px'}}
                                color={'#fff'}
                                icon='ios-create-outline'
                            />
                        </a>
                        <a className="col-1" role="button" onClick={(e) => {e.preventDefault(); onDeleteItem(item.id)}}>
                            <Ionicon
                                className="rounded"
                                fontSize="30px"
                                style={{backgroundColor: '#dc3545', padding:'5px'}}
                                color={'#fff'}
                                icon='ios-close'
                            />
                        </a>
                    </li>
                ))
            }
        </ul>
    )
}

PriceList.prototype = {
    item: PropTypes.array.isRequired,
    onMordifyItem: PropTypes.func.isRequired,
    onDeleteItem: PropTypes.func.isRequired
}

PriceList.defaultProps = {
    item: []
}

export default PriceList