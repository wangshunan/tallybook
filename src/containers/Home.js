import React from 'react';
import 'bootstrap/dist/css/bootstrap.min.css'
import Prilist from '../components/PriceList'
import ViewTab from '../components/ViewTab'
import {LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, parseToYearAndMonth, padLeft} from '../utility'
import TotalPrice from '../components/TotalPrice';
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'

export const categories = {
    "1": {
        "id": 1,
        "name": "旅行",
        "type": TYPE_OUTCOME,
        "iconName": "ios-plane"
    },
    "2": {
        "id": 2,
        "name": "旅行",
        "type": TYPE_OUTCOME,
        "iconName": "ios-plane"
    }
}

export const testItems = [
    {
      "id": 1,
      "title": "test",
      "price": 200,
      "date": "2020-07-10",
      "cid": 1
    },
    {
      "id": 2,
      "title": "test",
      "price": 200,
      "date": "2016-09-10",
      "cid": 2
    }
]

  class Home extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            items: testItems,
            currentDate:parseToYearAndMonth(),
            tabView: LIST_VIEW
        }
    }

    changeView = (view) => {
        this.setState({
            tabView: view
        })
    }

    changeDate = (date) => {
        this.setState({
            currentDate: date
        })
    }

    modifyItem = (id) => {
        const modifiedItems = this.state.items.map(item => {
            if ( id === item.id ) {
                return {...item, title:'test2'}
            } else {
                return item
            }
        })

        this.setState({
            items: modifiedItems
        })
    }

    createItem = () => {
        const newItem = {
            "id": 3,
            "title": "test",
            "price": 500,
            "date": "2018-08-10",
            "cid": 1
        }

        this.setState({
            items: [newItem, ...this.state.items]
        })
    }

    deleteItem = (id) => {
        const newItems = this.state.items.filter(item => item.id !== id)
        this.setState({
            items: newItems
        })
    }

    render() {
        let totalIncome = 0, totalOutcome = 0
        const { currentDate, items, tabView } = this.state
        const itemsWithCategory = items.map(item => {
            item.category = categories[item.cid]
            return item
        }).filter(item => {
            return item.date.includes(`${currentDate.year}-${padLeft(currentDate.month)}`)
        })
        itemsWithCategory.forEach(item => {
            if ( item.category.type === TYPE_INCOME ) {
                totalIncome += item.price
            } else {
                totalOutcome += item.price
            }
        })
        return (
            <React.Fragment>
                <header>
                </header>
                <div className="row home-container align-items-center">
                    <div className="col m-4">
                        <MonthPicker
                            year={currentDate.year}
                            month={currentDate.month}
                            onChange={this.changeDate}
                        />
                    </div>
                    <div className="col text-center">
                        <TotalPrice 
                            income={totalIncome}
                            outcome={totalOutcome}
                        />
                    </div>
                </div>
                <div className="content-area py-3 px-3">
                    <ViewTab
                        activeTab={tabView}
                        onTabChange={this.changeView}
                    />
                    <CreateBtn
                        btnName="创建一条新的记账记录"
                        iconName="md-add"
                        callBack={this.createItem}
                    />
                    { tabView === LIST_VIEW &&
                        <Prilist
                            items={itemsWithCategory}
                            onModifyItem={this.modifyItem}
                            onDeleteItem={this.deleteItem}
                        />
                    }
                    { tabView === CHART_VIEW &&
                        <h1>图表模式</h1>
                    }
                </div>
            </React.Fragment>
        )
    }
  }

  export default Home