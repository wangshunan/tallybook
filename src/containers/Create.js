import React from 'react'
import 'bootstrap/dist/css/bootstrap.min.css'
import CategorySelect from '../components/CategorySelect'
import { Tabs, Tab } from '../components/Tabs'
import PriceForm from '../components/PriceForm'
import WithContext from '../WithContext'
import { withRouter } from 'react-router-dom'
import { TYPE_OUTCOME, TYPE_INCOME } from '../utility'

const tabsText = [TYPE_OUTCOME, TYPE_INCOME]
class Create extends React.Component {
    constructor(props) {
        super(props)
        const { id } = props.match.params
        const { categories, items } = props.data
        this.state = {
            selectedTab: (id && items[id]) ? categories[items[id].cid].type : TYPE_OUTCOME,
            selectedCategory: ( id && items) ? categories[items[id].cid] : null
        }
    }

    tabChange = (index) => {
        this.setState({
            selectedTab: tabsText[index]
        })
    }

    submitFrom = (data, editMode) => {
        if (editMode) {
            // create
            this.props.actions.createItem(data, this.state.selectedCategory.id)
        } else {
            // update
            this.props.actions.updateItem(data, this.state.selectedCategory.id)
        }
        this.props.history.push('/')
    }

    cancelSubmit = () => {
        this.props.history.push('/')
    }

    render() {
        const { data } = this.props
        const { items, categories } = data
        const { selectedTab, selectedCategory } = this.state
        const { id } = this.props.match.params
        const editItem = (id && items[id]) ? items[id] : {}
        const filterCategories = Object.keys(categories)
        .map(id => {
            if (categories[id].type === selectedTab) {
                return categories[id]
            }})
        .filter((e) => {return e})
        const tabIndex = tabsText.findIndex(text => text === selectedTab)
        console.log(selectedCategory)

        return (
            <div className="container py-3 px-5 align-items-center">
                <Tabs activeIndex={tabIndex} onTabChange={this.tabChange}>
                    <Tab>
                        outcome
                    </Tab>
                    <Tab>
                        income
                    </Tab>
                </Tabs>
                <CategorySelect
                    categories={filterCategories}
                    onSelectCategory={(category) => {this.setState({selectedCategory: category})}}
                    selectedCategory={selectedCategory}
                />
                <PriceForm
                    item={editItem}
                    onFormSubmit={this.submitFrom}
                    onCancelSubmit={this.cancelSubmit}
                />
            </div>
        )
    }
}

export default withRouter(WithContext(Create))