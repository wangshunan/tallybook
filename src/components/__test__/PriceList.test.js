import React from 'react'
import { shallow } from 'enzyme'
import PriceList from '../PriceList'
import { testItems, categories } from '../../containers/Home'
import toJson from 'enzyme-to-json'
import Ionicon from 'react-ionicons'

const itemsWithCategory = testItems.map(item => {
    item.category = categories[item.cid]
    return item
})

const props = {
    items: itemsWithCategory,
    onModifyItem: jest.fn(),
    onDeleteItem: jest.fn()
}

let wrapper
describe('test PriceList component', () => {
    beforeEach(() => {
        wrapper = shallow(<PriceList {...props}/>)
    })

    it('should render the component to match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should render correct price items length', () => {
        expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
    })

    it('should render correct icon and price for each item', () => {
        const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
        expect(iconList.length).toEqual(3)
        expect(iconList.first().props().icon).toEqual(itemsWithCategory[0].category.iconName)
    })

    it('should trigger the correct function callbacks', () => {
        const firstItem = wrapper.find('.list-group-item').first()
        firstItem.find('a').first().simulate('click', { preventDefault: () => {} })
        expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0].id)

        firstItem.find('a').last().simulate('click', { preventDefault: () => {} })
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0].id)
    })
})