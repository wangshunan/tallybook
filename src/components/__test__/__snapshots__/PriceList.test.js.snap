// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshot 1`] = `
<ul
  className="list-group"
>
  <li
    className="list-group-item d-flex justify-content-between align-items-center row"
    key="1"
    style={
      Object {
        "margin": "0px",
      }
    }
  >
    <span
      className="col-1 badege badege-primary"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-6"
    >
      test
    </span>
    <span
      className="col-1"
    >
      -200
    </span>
    <span
      className="col-2"
    >
      2020-07-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center row"
    key="2"
    style={
      Object {
        "margin": "0px",
      }
    }
  >
    <span
      className="col-1 badege badege-primary"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-6"
    >
      test
    </span>
    <span
      className="col-1"
    >
      -200
    </span>
    <span
      className="col-2"
    >
      2016-09-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
    >
      <Ionicon
        beat={false}
        className="rounded"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
